daprblog/content/posts/2023/dapr_flomesh/index.md

```diff
@@ -0,0 +1,546 @@
+  echo '-----------------------------------'
+  kubectl get serviceexports.flomesh.io -A
+  echo '-----------------------------------'
+  curl -s "http://${HOST_IP}:${PORT}/ports"
+  echo '-----------------------------------'
+done
+```
+
+## Test
+
+Switch to cluster `cluster-2` and make a request from the `curl` pod to test.
+
+```shell
+kubectx k3d-cluster-2
+curl_client="$(kubectl get pod -n curl -l app=curl -o jsonpath='{.items[0].metadata.name}')"
+```
+
+An error occurred when sending a request to access the nodeapp because the nodeapp application was not deployed in `cluster-2`. By default, when no cross-cluster traffic policy is specified, it will only attempt to call local services and will not schedule traffic to other clusters.
+
+```shell
+kubectl exec "${curl_client}" -n curl -c curl -- curl -s http://nodeapp.dapr-test:3000/ports
+command terminated with exit code 7
+```
+
+### Setting Traffic Policies
+
+There are three types of traffic policies for cross-cluster traffic management:
+
+- `Locality`: Only use local services, which is the default type. This is why accessing the nodeapp application fails when we don't provide any global policies, because the service is not available in the `cluster-2`.
+- `FailOver`: Only proxy to other clusters when local cluster fails, similar to the concept of active-passive failover.
+- `ActiveActive`: Proxy to other clusters normally even when local cluster is up and running, similar to active-active.
+
+
+Next, we will create and apply an `ActiveActive` policy in the `cluster-2` cluster.
+
+```shell
+kubectl apply -n dapr-test -f - <<EOF
+apiVersion: flomesh.io/v1alpha1
+kind: GlobalTrafficPolicy
+metadata:
+  name: nodeapp
+spec:
+  lbType: ActiveActive
+  targets:
+    - clusterKey: default/default/default/cluster-1
+      weight: 100
+    - clusterKey: default/default/default/cluster-3
+      weight: 100
+EOF
+```
+
+Sending the request again now results in a successful response.
+
+```shell
+kubectl exec "${curl_client}" -n curl -c curl -- curl -s http://nodeapp.dapr-test:3000/ports
+{"DAPR_HTTP_PORT":"3500","DAPR_GRPC_PORT":"50001"}, from cluster: cluster-3
+```
+
+This time, try to send a request to `/neworder` to attempt to write order data.
+
+```shell
+kubectl exec "${curl_client}" -n curl -c curl -- curl -si --request POST --data '{"data":{"orderId":"42"}}' --header Content-Type:application/json --header dapr-app-id:nodeapp http://nodeapp.dapr-test:3000/neworder
+created order via cluster: cluster-1
+```
+
+Multiple requests to `/order` can be made, and it can be observed that the requests are forwarded to different clusters for processing.
+
+```shell
+kubectl exec "${curl_client}" -n curl -c curl -- curl -s http://nodeapp.dapr-test:3000/order
+{"orderId":"42"}, from cluster: cluster-3
+kubectl exec "${curl_client}" -n curl -c curl -- curl -s http://nodeapp.dapr-test:3000/order
+{"orderId":"42"}, from cluster: cluster-1
+```
+
+
```

Comment:
Maybe add a quick paragraph to conclude - drive home the benefit of Flomesh and Dapr together?